src/provider/mlkzg.rs

```diff
@@ -0,0 +1,621 @@
+                          transcript: &mut <NE as NovaEngine>::TE|
+     -> (Vec<E::G1Affine>, Vec<Vec<E::Fr>>) {
+      let poly_eval = |f: &[E::Fr], u: E::Fr| -> E::Fr {
+        // Horner's scheme
+        let mut v = f[f.len() - 1];
+
+        for fi in f.iter().rev().skip(1) {
+          v *= u;
+          v += fi;
+        }
+
+        v
+      };
+
+      let scalar_vector_muladd = |a: &mut Vec<E::Fr>, v: &Vec<E::Fr>, s: E::Fr| {
+        assert!(a.len() >= v.len());
+        for i in 0..v.len() {
+          a[i] += s * v[i];
+        }
+      };
+
+      let kzg_compute_batch_polynomial = |f: &[Vec<E::Fr>], q: E::Fr| -> Vec<E::Fr> {
+        let k = f.len(); // Number of polynomials we're batching
+
+        let q_powers = Self::batch_challenge_powers(q, k);
+
+        // Compute B(x) = f[0] + q*f[1] + q^2 * f[2] + ... q^(k-1) * f[k-1]
+        let mut B = f[0].clone();
+        for i in 1..k {
+          scalar_vector_muladd(&mut B, &f[i], q_powers[i]); // B += q_powers[i] * f[i]
+        }
+
+        B
+      };
+      ///////// END kzg_open_batch closure helpers
+
+      let k = f.len();
+      let t = u.len();
+      assert!(C.len() == k);
+
+      // The verifier needs f_i(u_j), so we compute them here
+      // (V will compute B(u_j) itself)
+      let mut v = vec![vec!(E::Fr::ZERO; k); t];
+      for i in 0..t {
+        // for each point u
+        for (j, f_j) in f.iter().enumerate().take(k) {
+          // for each poly f
+          v[i][j] = poly_eval(f_j, u[i]); // = f_j(u_i)
+        }
+      }
+
+      let q = Self::get_batch_challenge(C, u, &v, transcript);
+      let B = kzg_compute_batch_polynomial(f, q);
+
+      // Now open B at u0, ..., u_{t-1}
+      let mut w = Vec::with_capacity(t);
+      for ui in u {
+        let wi = kzg_open(&B, *ui);
+        w.push(wi);
+      }
+
+      // Compute the commitment to the batched polynomial B(X)
+      let q_powers = Self::batch_challenge_powers(q, k);
+      let c_0: E::G1 = C[0].into();
+      let C_B = (c_0 + NE::GE::vartime_multiscalar_mul(&q_powers[1..k], &C[1..k])).preprocessed();
+
+      // The prover computes the challenge to keep the transcript in the same
+      // state as that of the verifier
+      let _d_0 = Self::verifier_second_challenge(&C_B, &w, transcript);
+
+      (w, v)
+    };
+
+    ///// END helper closures //////////
+
+    let ell = x.len();
+    let n = hat_P.len();
+    assert_eq!(n, 1 << ell); // Below we assume that n is a power of two
+
+    // Phase 1  -- create commitments com_1, ..., com_\ell
+    let mut polys: Vec<Vec<E::Fr>> = Vec::new();
+    polys.push(hat_P.to_vec());
+    for i in 0..ell {
+      let Pi_len = polys[i].len() / 2;
+      let mut Pi = vec![E::Fr::ZERO; Pi_len];
+
+      #[allow(clippy::needless_range_loop)]
+      for j in 0..Pi_len {
+        Pi[j] = x[ell-i-1] * polys[i][2*j + 1]            // Odd part of P^(i-1)
+                      + (E::Fr::ONE - x[ell-i-1]) * polys[i][2*j]; // Even part of P^(i-1)
+      }
```

Comment:
The `Pi`s need to be computed sequentially, but their derivation can be parallelized.

Reply:
Fixed in https://github.com/lurk-lab/arecibo/pull/233